File: Leetcode/cpp/759.cpp
#include <iostream>
#include <vector>

/*
// Definition for an Interval.
class Interval {
public:
    int start;
    int end;

    Interval() {}

    Interval(int _start, int _end) {
        start = _start;
        end = _end;
    }
};
*/

class Solution {
public:
    std::vector<Interval> employeeFreeTime(std::vector<std::vector<Interval>> schedule) {
        std::vector<std::pair<int, int>> data;
        for (auto times : schedule) {
            for (auto time : times) {
                data.push_back({time.start, 1});
                data.push_back({time.end, -1});
            }
        }
        std::sort(data.begin(), data.end(), [](const std::pair<int, int>& a, const std::pair<int, int>& b) {
            if (a.first != b.first) return a.first < b.first;
            return a.second > b.second;
        });
        std::vector<Interval> output;
        int last = -1;
        int total = 0;
        for (auto d : data) {
            total += d.second;
            if (last != -1) {
                Interval interv;
                interv.start = last;
                interv.end = d.first;
                output.push_back(interv);
                last = -1;
            } else if (total == 0 && last == -1) {
                last = d.first;
            }
        }
        return output;
    }
};

File: Leetcode/Python/759.py
"""
# Definition for an Interval.
class Interval:
    def __init__(self, start: int = None, end: int = None):
        self.start = start
        self.end = end
"""

class Solution:
    def employeeFreeTime(self, schedule: '[[Interval]]') -> '[Interval]':
        schedule_time = defaultdict(int)
        flat_schedule = []

        for intervals in schedule:
            for interval in intervals:
                start = interval.start
                end = interval.end
                flat_schedule.append((start, 1))
                flat_schedule.append((end, -1))

        flat_schedule.sort(key=lambda x: (x[0], -x[1]))
        employee = 0
        for time, emp in flat_schedule:
            employee += emp
            schedule_time[time] = employee

        output = []
        last = -1
        for key in sorted(schedule_time.keys()):
            if last != -1:
                interv = Interval(last, key)
                output.append(interv)
                last = -1
            elif schedule_time[key] == 0:
                last = key
        return output
